File: sol/solution/0101-0200/0105/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Map<Integer, Integer> inorderMap = new HashMap<>();
    int idx = 0;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i = 0; i < inorder.length; i++)
            inorderMap.put(inorder[i], i);
        return helper(preorder, 0, inorder.length - 1);
    }
    private TreeNode helper(int[] preorder, int left, int right) {
        if (left > right) return null;
        int rootVal = preorder[idx++];
        TreeNode root = new TreeNode(rootVal);
        root.left = helper(preorder, left, inorderMap.get(rootVal) - 1);
        root.right = helper(preorder, inorderMap.get(rootVal) + 1, right);
        return root;
    }
}

File: sol/solution/0101-0200/0105/Solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private var index = 0
    private lateinit var inorderMap: Map<Int, Int>
    fun buildTree(preorder: IntArray, inorder: IntArray): TreeNode? {
        inorderMap = inorder.withIndex().associate { it.value to it.index }
        return helper(preorder, 0, inorder.size - 1)
    }
    private fun helper(preorder: IntArray, left: Int, right: Int): TreeNode? {
        if (left > right) return null
        val rootVal = preorder[index++]
        val root = TreeNode(rootVal)
        root.left = helper(preorder, left, inorderMap[rootVal]!! - 1)
        root.right = helper(preorder, inorderMap[rootVal]!! + 1, right)
        return root
    }
}
